cmd/root/exercise.go

```diff
@@ -49,30 +45,155 @@ func (m exerciseModel) renderName() string {
+		windowStart = currLine - linesBefore
+		windowEnd = windowStart + windowSize
+	case currLine >= len(lines)-linesAfter:
+		windowEnd = len(lines)
+		windowStart = windowEnd - windowSize
+	}
 
-	typed := m.typedText
+	// should show the whole exercise
+	if windowSize == 0 {
+		windowStart = 0
+		windowEnd = len(lines)
+	}
+
+	vignetteLastLine := true
+	if windowEnd == len(lines) {
+		vignetteLastLine = false
+	}
+
+	for i := windowStart; i < windowEnd; i = i + 1 {
+		text := lines[i]
+		var typed *string = nil
+		if i < len(typedLines) {
+			typed = &typedLines[i]
+		}
+		isCurrLine := i == currLine
+		shouldVignette := false
+		if vignetteLastLine && i == windowEnd-1 && i != windowStart {
+			shouldVignette = true
+		}
+		line := renderLine(text, typed, m.viewOptions.styles, shouldVignette, isCurrLine)
+		if lastLine := i == windowEnd-1; lastLine {
+			line = removeLastNewline(line)
+		}
+		s += line
+	}
+	return
+}
+
+// Splits up a string of text by newlines.
+// The newlines are preserved, since they'll be used
+// in rendering, too.
+func lines(text string) []string {
+	arr := strings.SplitAfter(text, "\n")
+	if last := len(arr) - 1; arr[last] == "" {
+		arr = arr[:last]
+	}
+	return arr
+}
+
+// Returns an array of strings that map the typed characters
+// to the exercise characters. If no characters have been typed
+// on a current line, the typedLine will be nil.
+func typedLines(lines []string, typed string) []string {
+	typedLines := []string{}
+	i := 0
+	for _, line := range lines {
+		str := ""
+		for range line {
+			if i >= len(typed) {
+				continue
+			}
+			str = str + string(typed[i])
+			i = i + 1
+		}
+		if str != "" {
+			typedLines = append(typedLines, str)
+		}
+	}
+	return typedLines
+}
+
+func currentLineI(lines []string, typed string) int {
+	typedLen := len(typed)
+	for i := range lines {
+		for range lines[i] {
+			if typedLen == 0 {
+				return i
+			}
+			typedLen = typedLen - 1
+		}
+	}
+	return 0
+}
+
+func removeLastNewline(str string) string {
+	n := '\n'
+	i := len(str) - 1
+	for ; i >= 0 && rune(str[i]) != n; i = i - 1 {
+	}
+
+	if i < 0 {
+		return str
+	}
+
+	return fmt.Sprintf("%s%s", str[:i], str[i+1:])
+}
+
+func renderLine(text string, typedP *string, style styles, vignette bool, currLine bool) (s string) {
+	typedStyle := style.typedStyle
+	untypedStyle := style.untypedStyle
+	cursorStyle := style.cursorStyle
+	mistakeStyle := style.mistakeStyle
+
+	if vignette {
+		typedStyle = style.vignetteStyle
+		untypedStyle = style.vignetteStyle
+		cursorStyle = style.vignetteStyle
+		mistakeStyle = style.vignetteMistakeStyle
+	}
 
-	for i, exRune := range m.text {
+	if typedP == nil {
+		for i, c := range text {
+			currChar := string(c)
+			if c != '\n' {
+				currChar = untypedStyle.Render(string(c))
+			}
+			if i == 0 && currLine {
+				currChar = renderVisibleRune(cursorStyle, c)
+			}
+			s += currChar
+		}
+		return
+	}
+
+	typed := *typedP
+
+	for i, exRune := range text {
```

Comment:
I could probably just redo this block as a expressionless switch statement, which may make things easier to read.